subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/annotations/RunFor.groovy

```diff
@@ -0,0 +1,148 @@
+            if (SCENARIO_DEFINITION_FILE != null) {
+                if (WRITE_SCENARIO_JSON) {
+                    scenarioDefinition.writeTo(SCENARIO_DEFINITION_FILE)
+                } else {
+                    scenarioDefinition.verify(SCENARIO_DEFINITION_FILE)
+                }
+            }
+        } as Runnable))
+    }
+
+    @Override
+    void visitSpecAnnotation(RunFor runFor, SpecInfo spec) {
+        assert runFor.value().every { it.iterationMatcher().isEmpty() }: "No iterationMatchers allowed in class-level @Scenario!"
+        for (FeatureInfo feature : spec.getFeatures()) {
+            if (!feature.getFeatureMethod().getReflection().isAnnotationPresent(RunFor.class)) {
+                visitFeatureAnnotation(runFor, feature)
+            }
+        }
+    }
+
+    @Override
+    void visitFeatureAnnotation(RunFor runFor, FeatureInfo feature) {
+        feature.getFeatureMethod().addInterceptor(new AddScenarioDefinitionInterceptor(feature.getFeatureMethod().getReflection(), runFor))
+    }
+
+    private static void verify(Scenario scenario) {
+        assert scenario.testProjects().length != 0 && scenario.operatingSystems().length != 0: "Invalid scenario: $scenario"
+    }
+
+    private static GroupsBean createGroup(Scenario scenario, String testProject) {
+        return new GroupsBean(
+            testProject: testProject,
+            coverage: [(scenario.type().toString().toLowerCase()): scenario.operatingSystems().collect { it.coverageName }] as TreeMap,
+            comment: scenario.comment() ?: null
+        )
+    }
+
+    static class AddScenarioDefinitionInterceptor implements IMethodInterceptor {
+        private final Method method
+        private final RunFor runFor
+
+        AddScenarioDefinitionInterceptor(Method method, RunFor runFor) {
+            this.method = method
+            this.runFor = runFor
+        }
+
+        @Override
+        void intercept(IMethodInvocation invocation) throws Throwable {
+            if (SCENARIO_DEFINITION_FILE != null) {
+                String testId = invocation.getIteration().getName()
+
+                List<Scenario> scenarios = runFor.value().toList()
+                List<GroupsBean> groups = []
+                for (Scenario scenario : scenarios) {
+                    verify(scenario)
+                    String matcher = scenario.iterationMatcher()
+                    if (matcher.isEmpty() || testId.matches(matcher)) {
+                        scenario.testProjects().each {
+                            groups.add(createGroup(scenario, it))
+                        }
+                    }
+                }
+                if (!groups.isEmpty()) {
+                    scenarioDefinition.getPerformanceTests().add(new PerformanceTestScenarioDefinition.PerformanceTestsBean("${invocation.getSpec().getReflection().getName()}.$testId", groups))
+                }
+                Assume.assumeFalse(true)
+            }
```

Comment:
```suggestion
            } else {
                invocation.proceed()
            }
```

This seems to work. https://ge.gradle.org/s/bg3khh6zlugqk and https://ge.gradle.org/s/s4uojtkjb4o3s/tests